feat: Add HOWTO: Create & test a TCP load balancer (#116)
citynetwork/docs: docs/howto/openstack/octavia/lbaas-tcp.md

steps are presented and explained through a specific scenario, and we
work using either the {{gui}} or the OpenStack CLI.

## Prerequisites

Whether you choose the {{gui}} or the OpenStack CLI, you need to [have
an account](/howto/getting-started/create-account) in {{brand}}.
Additionally, to use the OpenStack CLI, make sure to [enable
it](/howto/getting-started/enable-openstack-cli) for the region you
will be working in. Besides the Python `openstackclient` module, you
will also have to install the Python `octaviaclient` module. For that,
use either the package manager of your operating system or `pip`:

=== "Debian/Ubuntu"
    ```bash
    apt install python3-octaviaclient
    ```
=== "Mac OS X with Homebrew"
    This particular Python module is unavailable via `brew`, but you
    can install it via `pip`.
=== "Python Package"
    ```bash
    pip install python-octaviaclient
    ```

## Scenario and terminology

To configure and test our basic TCP load balancer, we have two servers
in the same {{brand}} region, both having `ncat` listening on port
61234/TCP and returning a greeting to clients asking to connect to that

> port. Both servers are in the same internal network, and our load

Author: Hmm, while it is correct to say, "I am on Facebook", "I'm on the network", etc., when we are talking about _things_ (not humans) that are actual parts of a network, then we use "in". This is my understanding --- and I am sure I often get this wrong.

balancer will be responsible for redirecting client connections to them
in a round-robin fashion. The load balancer will have a single floating
IP address, so prospective clients will have to know that IP and not

> any of the backend servers.

Author: Yes, I will update accordingly.

    To confirm the changes, click the green _Update_ button.

    ![Apply changes to pool](assets/shot-17.png)

    At this point, the load balancer should have its listener plus a pool
    with two members.

    ![Pool is populated](assets/shot-18.png)
=== "OpenStack CLI"
    Both our test servers are in the `subnet-fra1` subnet, one of them has
    IP `10.15.25.105`, and the other one has IP `10.15.25.236`. To add
    those two servers in pool `mylb-pool`, type:

    ```bash
    openstack loadbalancer member create \
        --subnet-id subnet-fra1 --address 10.15.25.105 \
        --protocol-port 61234 mylb-pool
    ```

    ```text
    +---------------------+--------------------------------------+
    | Field               | Value                                |
    +---------------------+--------------------------------------+
    | address             | 10.15.25.105                         |
    | admin_state_up      | True                                 |
    | created_at          | 2023-01-16T20:11:08                  |
    | id                  | 93aed0c8-b425-40b5-b166-f3e44ad9a986 |
    | name                |                                      |

>     | operating_status    | NO_MONITOR                           |

Author: Ah, yes, I did not mention health monitoring at all. I will make sure to update the guide accordingly.

    | tags                | []                                   |
    | updated_at          | 2023-01-16T20:25:25Z                 |
    +---------------------+--------------------------------------+
    ```

    Then, load the ID of the new floating IP to a variable:

    ```bash
    FLOAT_IP_ID=$(openstack floating ip list \
        --floating-ip-address 198.51.100.129 -c ID -f value)
    ```

    Do the same for the VIP port ID of your load balancer:

    ```bash
    VIP_PORT_ID=$(openstack loadbalancer show mylb \
        -c vip_port_id -f value)
    ```

    Finally, assign the floating IP to your load balancer like so:

    ```bash
    openstack floating ip set --port $VIP_PORT_ID $FLOAT_IP_ID
    ```

## Testing the load balancer

Each of the two test servers runs `ncat` to continuously listen
on port 61234/TCP and respond to clients with a simple message,
revealing its hostname:

```bash
ncat -kv -l 61234 -c 'echo Yello from $(hostname)!'
```

Since the load balancer has a floating IP (`198.51.100.129`,
in our example). and we know the port it listens to, we can
try connecting to it via `wget` and see what happens:

```bash

> wget -q 198.51.100.129:61234 -O -

Author: Indeed, `curl` looks like the better choice. In our case, though, we need to add the `--http0.9` parameter. If we don't, we get this error: `curl: (1) Received HTTP/0.9 when not allowed`. While I am fine with adding the `--http0.9` option to `curl`, I suspect that the `wget` route feels more intuitive/natural/straightforward to the reader. What do you think? Should we go with `wget` or with `curl`?

> wget -q 198.51.100.129:61234 -O -

Reviewer: I think you can use `curl` with a `telnet://` URL, no? I mean, you're not _actually_ doing HTTP here anywhere, so why pretend that you do?

> wget -q 198.51.100.129:61234 -O -

Author: I just tried with `telnet://` and it doesn't work. So, as fae as I can tell, it's going to be either `wget -q p.q.r.xyz:61234 -O -` or `curl --http0.9 p.q.r.xyz:61234`.

> wget -q 198.51.100.129:61234 -O -

Reviewer: I just followed the steps on my local machine and then ran `curl -s telnet://localhost:61234` and got the expected response back...

> wget -q 198.51.100.129:61234 -O -

Author: Interesting. From Mac OS X, it doesn't work (you get nothing back). From a Raspberry Pi with Debian 11 it *does* work, but you have to hit [CTRL+C] to get your command line back. Let us use something that works on both platforms, meaning either `wget` or `curl` (with no `telnet://`). What do you think?

> wget -q 198.51.100.129:61234 -O -

Reviewer: OK, let's use `wget` then. Apologies for the noise. :)

> wget -q 198.51.100.129:61234 -O -

Author: `wget` it is, then --- and no noise at all!

```

```plain
Yello from srv-lbaas-1!
```

It looks like we talked to the first test server. If we run the
exact same `wget` command for a second time, since the load balancer
distributes client connection requests in a round-robin fashion,
we expect to talk to the second server:

```bash
wget -q 198.51.100.129:61234 -O -
```

```plain
Yello from srv-lbaas-2!
```

During our testing, executing the `wget` command repeatedly we were
getting those two hostnames one after the other --- and that was

> proof our TCP load balancer was working as expected.

Author: Aha! That would greatly expand our test coverage. I will make sure to update the corresponding section accordingly.

    | revision_number     | 0                                    |
    | router_id           | None                                 |
    | status              | DOWN                                 |
    | subnet_id           | None                                 |
    | tags                | []                                   |
    | updated_at          | 2023-01-16T20:25:25Z                 |
    +---------------------+--------------------------------------+
    ```

    Then, load the ID of the new floating IP to a variable:

    ```bash
    FLOAT_IP_ID=$(openstack floating ip list \
        --floating-ip-address 198.51.100.129 -c ID -f value)
    ```

    Do the same for the VIP port ID of your load balancer:

    ```bash
    VIP_PORT_ID=$(openstack loadbalancer show mylb \
        -c vip_port_id -f value)
    ```

    Finally, assign the floating IP to your load balancer like so:

    ```bash
    openstack floating ip set --port $VIP_PORT_ID $FLOAT_IP_ID
    ```

## Testing the load balancer

Each of the two test servers runs `ncat` to continuously listen
on port 61234/TCP and respond to clients with a simple message,
revealing its hostname:

```bash

> ncat -kV -l 61234 -c 'echo Yello from $(hostname)!'

Reviewer: I think this may have been accidentally changed in 4698e923cdb46a6c604cd9ffb706a8305b1742a8. `-kv` is working for me, `-kV` is not.

Author: That was accidental (and nice catch, BTW).

```

Since the load balancer has a floating IP (`198.51.100.129`,

> in our example). and we know the port it listens to, we can

Reviewer: Looks like a full-stop slipped in here instead of a comma; sorry for not catching this earlier.

```suggestion
in our example), and we know the port it listens to, we can
```

Author: **I'm** sorry for not seeing that :/

try connecting to it via `wget` and see what happens:

```bash
wget -q 198.51.100.129:61234 -O -
```

```plain
Yello from srv-lbaas-1!
```

It looks like we talked to the first test server. If we run the
exact same `wget` command for a second time, since the load balancer
distributes client connection requests in a round-robin fashion,
we expect to talk to the second server:

```bash
wget -q 198.51.100.129:61234 -O -
```

```plain
Yello from srv-lbaas-2!
```

During our testing, executing the `wget` command repeatedly, we were
getting those two hostnames one after the other --- and that was
proof our TCP load balancer was working as expected. But there is

> one more expectation of any load balancer; the ability to skip

Reviewer: I know I am being an extreme stickler here, but the semicolon is incorrect (since the clause coming after is has an object, but is lacking a subject and a predicate). :) 

```suggestion
one more expectation of any load balancer: the ability to skip
```

Author: Oops, I admit I am not aware of colon/semicolon-related rules :/

backend hosts when they are inaccessible.

To test our load balancer in this new scenario, let us first use
`wget` to connect to port 61234 and jot down the backend server
that will respond:

```bash
wget -q 198.51.100.129:61234 -O -
```

```plain
Yello from srv-lbaas-2!
```

We see that `srv-lbaas-2` responded. That means the next time we

> try to connect, `srv-lbaas-1` will respond. But if we SSH into

Reviewer: Residual reference to SSH.

Author: No, it is not, I am just describing a course of action.

`srv-lbaas-1` and terminate `ncat`, then after connecting with
`wget` we will get a response from `srv-lbaas-2` --- again. This,

> at least, is our expectation. So without further ado, we SSH into

Reviewer: Another residual reference to SSH.

Author: Again, it is not :)

You can add a health monitor to the pool of your load balancer,
so whenever ---and for whatever reason--- one or more of the pool
members are inaccessible, the load balancer will know and won't
even bother trying to redirect client requests to them. Of course,
whenever an inaccessible pool member gets accessible again, the
load balancer will take notice and start treating that member as
a fully functional server.

=== "{{gui}}"
    In the detailed view of your load balancer, pull up the
    _Pools_ tab. In the _Health Monitor_ column there is a "0",
    for the pool has no health monitor yet.

    ![There is no health monitor](assets/shot-23.png)
    
    To add a health monitor, click the notepad-and-pen icon.
    A pane titled _Modify Pool_ slides over. Scroll down if
    you have to, and stop when the _Health Monitor_ section is
    fully visible. You will see a message saying, "No health
    monitor created", so click the green _Create a Healthmonitor_
    button to create one.
    
    ![Create new health monitor](assets/shot-24.png)
    
    Pick a name for the new health monitor, and set its type to
    PING. That way, ICMP echo requests will be used to determine

>     whether pool members are accessible.

Reviewer: Let's please not do that; that's another bad practice we'd prefer users not to do. Whether a _host_ is pingable really has no bearing on whether the _service_ is accessible. If the host is up but the service has been stopped, we want that host to be dropped from the pool, and when we use a `PING` health monitor we don't. So let's use a `TCP` one instead.

Author: I see --- and I will update accordingly!

    | 76ddfdc73 |           |            |                     |           |               |                  |        |
    | ca2       |           |            |                     |           |               |                  |        |
    | 04a32731- | srv-      | dfc7004673 | ACTIVE              | 10.15.25. |         61234 | ONLINE           |      1 |
    | f3a6-     | lbaas-2   | 96428bacba |                     | 236       |               |                  |        |
    | 41ca-     |           | 4376e72cc3 |                     |           |               |                  |        |
    | a638-     |           | e9         |                     |           |               |                  |        |
    | 6f8767437 |           |            |                     |           |               |                  |        |
    | 50c       |           |            |                     |           |               |                  |        |
    +-----------+-----------+------------+---------------------+-----------+---------------+------------------+--------+

>     ```

Reviewer: There's one more note I would include here (because it is frequently a source of confusion), which is what `openstack loadbalancer member list` looks like after you kill `ncat` on one server, and what your `wget` command does under the same circumstances (namely, that it hits only the remaining healthy server, rather than alternately hitting a healthy and a broken one).

Author: Ah, yes, that note would be useful. Adding and updating in a bit...

Reviewer: Ack, and in the process please also fix [the Markdown lint errors](https://github.com/citynetwork/docs/actions/runs/3986415488/jobs/6834966865#step:6:33) that the CI runs highlight.

Author: Yeah, although I run `tox -e markdownlint` before updating, sometimes I accidentally open the file from the *wrong* editor, errors creep in, and of course, I forget to re-run the linter :/

    | a638-     |           | e9         |                     |           |               |                  |        |
    | 6f8767437 |           |            |                     |           |               |                  |        |
    | 50c       |           |            |                     |           |               |                  |        |
    +-----------+-----------+------------+---------------------+-----------+---------------+------------------+--------+
    ```

    In the example above, see the `operating_status` column,
    specifically the status of server `srv-lbaas-1`. Regarding
    the operating status of all pool members, you can always
    limit the scope of your query like so:

    ```bash
    openstack loadbalancer member list mylb-pool \
        -c name -c operating_status
    ```

    ```plain
    +-------------+------------------+
    | name        | operating_status |
    +-------------+------------------+
    | srv-lbaas-1 | ERROR            |
    | srv-lbaas-2 | ONLINE           |
    +-------------+------------------+
    ```

> Finally, let us check what happens when we try to access the
> remote service via the load balancer:
> 
> ```bash
> wget -q 198.51.100.129:61234 -O -
> ```
> 
> ```plain
> Yello from srv-lbaas-2!
> ```
> 

Reviewer: I think you can drive the point home even better if you instruct the learner to run the command three times, like so: (note that the GitHub web interface chops off the closing code fence here; you'll have to re-instate it.)

```suggestion
Finally, let us check what happens when we try to access the
remote service via the load balancer, multiple times:

```console
$ wget -q 198.51.100.129:61234 -O -
Yello from srv-lbaas-2!

$ wget -q 198.51.100.129:61234 -O -
Yello from srv-lbaas-2!

$ wget -q 198.51.100.129:61234 -O -
Yello from srv-lbaas-2!
```

```

Author: Sure, I made some changes taking into account your suggestion!
